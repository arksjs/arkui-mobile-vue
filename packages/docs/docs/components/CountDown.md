# CountDown 倒计时

注：

- 本组件没有指定样式，默认情况下可对文字样式进行自定义。

<CodeDemo name="CountDown">

<<< @/../../demo/src/components/Show/CountDown/index.vue

</CodeDemo>

## Import

```js
import { CountDown } from 'vfox'
```

具体的引入方式可以参考[引入组件](../guide/import.md)。

### Import Type

组件导出的类型定义：

```ts
import type {
  CountTime,
  CountDownOnEnd,
  CountDownOnPause,
  CountDownOnResume,
  CountDownRef
} from 'rfox'
```

## Props

| 属性           | 类型    | 默认值 | 必填 | 说明                |
| -------------- | ------- | ------ | ---- | ------------------- |
| initial-timing | number  | 99     | 否   | 倒计时时长，单位 ms |
| show-days      | boolean | false  | 否   | 是否显示天数        |

## Events

| 事件   | 描述           | 回调函数参数                                                                                              | TypeScript 函数   |
| ------ | -------------- | --------------------------------------------------------------------------------------------------------- | ----------------- |
| pause  | 计时暂停时触发 | payload: { remainTime: number } remainTime 剩余时间，单位 ms                                              | CountDownOnPause  |
| resume | 恢复计时时触发 | payload: { remainTime: number } remainTime 剩余时间，单位 ms                                              | CountDownOnResume |
| end    | 计时结束时触发 | payload: { startTime: number, endTime: number } startTime 本地开始时间戳，endTime 本地结束时间戳，单位 ms | CountDownOnEnd    |

## Slots

支持自定义风格显示

```vue
<ak-count-down :initial-timing="timestamp">
  <template #default="countTime">
    {{ countTime.fullHours }}:{{ countTime.minutes }}:{{
        countTime.seconds
    }}.{{ countTime.milliseconds }}
  </template>
</ak-count-down>
```

### countTime 的结构

| 字段名             | 类型   | 说明                                        |
| ------------------ | ------ | ------------------------------------------- |
| time               | number | 持续时间                                    |
| days               | string | 天数                                        |
| hours              | string | 小时数（<24），需要跟 days 配合，保留 2 位  |
| fullHours          | string | 小时数，含天数综合，如 '124'，至少保留 2 位 |
| thousandsFullHours | string | 千分位形式的小时数，含天数综合，如 '1,234'  |
| minutes            | string | 分钟数，保留 2 位                           |
| seconds            | string | 秒钟数，保留 2 位                           |
| milliseconds       | string | 毫秒数，保留 3 位                           |

## Methods

| 方法名 | 说明       | 参数                                            |
| ------ | ---------- | ----------------------------------------------- |
| pause  | 暂停倒计时 |                                                 |
| resume | 恢复倒计时 |                                                 |
| reset  | 重置倒计时 | ( timing: number, autoStart?: boolean ) 单位 ms |
