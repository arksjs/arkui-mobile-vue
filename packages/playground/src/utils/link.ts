export const genUnpkgLink = (
  pkg: string,
  version: string | undefined,
  path: string
) => {
  version = version ? `@${version}` : ''
  return `https://unpkg.com/${pkg}${version}${path}`
}
