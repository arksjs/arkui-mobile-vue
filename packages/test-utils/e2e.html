<script src="../../node_modules/vue/dist/vue.global.js"></script>
<script src="../../dist/index.js"></script>
<link rel="stylesheet" href="../../dist/index.css" />

<div id="app"></div>
