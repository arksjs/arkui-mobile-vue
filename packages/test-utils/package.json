{
  "name": "@arksjs/test-utils",
  "version": "0.0.1",
  "description": "arkui test-utils",
  "main": "dist/index.js",
  "private": true
}
