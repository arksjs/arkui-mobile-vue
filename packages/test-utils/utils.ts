export const timeout = (n: number) => new Promise(r => setTimeout(r, n))
