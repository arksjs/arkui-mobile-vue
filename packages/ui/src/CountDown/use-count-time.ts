import { onBeforeUnmount, ref } from 'vue'
import { getDefaultCountTime, getCountTime } from './util'

interface StepHandlers {
  update: (time: number) => void
  start: () => void
  stop: () => void
}

export function useCountTime(onStep: (handlers: StepHandlers) => void) {
  const times = ref(getDefaultCountTime())

  function update(time: number) {
    times.value = getCountTime(time)
  }

  let timer: number

  function start() {
    stop()
    timer = requestAnimationFrame(() => {
      start()
      onStep({ update, start, stop })
    })
  }

  function stop() {
    cancelAnimationFrame(timer)
  }

  onBeforeUnmount(stop)

  return {
    times,
    timeStart: start,
    timeStop: stop,
    timeUpdate: update
  }
}
