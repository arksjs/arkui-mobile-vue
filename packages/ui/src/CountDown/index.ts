import { withInstall } from '../helpers/with-install'
import CountDown from './CountDown.vue'

export { CountDown }
export const install = withInstall(CountDown)
export default CountDown
