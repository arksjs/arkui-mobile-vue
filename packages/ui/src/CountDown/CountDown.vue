<template>
  <div class="ta-count-down">
    <slot
      :time="countTime.time"
      :days="countTime.days"
      :fullHours="countTime.fullHours"
      :thousandsFullHours="countTime.thousandsFullHours"
      :hours="countTime.hours"
      :minutes="countTime.minutes"
      :seconds="countTime.seconds"
      :milliseconds="countTime.milliseconds"
    >
      {{
        showDays
          ? countTime.days + locale.countDownDayUnit + ' ' + countTime.hours
          : countTime.fullHours
      }}:{{ countTime.minutes }}:{{ countTime.seconds }}
    </slot>
  </div>
</template>

<script lang="ts">
import { defineComponent, onMounted } from 'vue'
import { useCountTime } from '../CountDown/use-count-time'
import { useLocale } from '../ConfigProvider/context'
import type { CountDownEmits, OnPauseOrResume, Reset } from './types'
import type { VoidFnToBooleanFn, PropsToEmits } from '../helpers/types'
import { isNumber } from '../helpers/util'

const pauseOrResumeValidator: VoidFnToBooleanFn<OnPauseOrResume> = payload =>
  payload && typeof payload.remainTime === 'number'

export default defineComponent({
  name: 'ta-count-down',
  props: {
    // 初始倒计时时间
    initialTiming: {
      type: Number,
      default: 0
    },
    // 是否显示天数
    showDays: {
      type: Boolean,
      default: false
    },
    // 是否暂停
    paused: {
      type: Boolean,
      default: false
    }
  },
  emits: {
    end: payload =>
      payload && isNumber(payload.startTime) && isNumber(payload.endTime),
    pause: pauseOrResumeValidator,
    resume: pauseOrResumeValidator
  } as PropsToEmits<CountDownEmits>,
  setup(props, { emit, expose }) {
    const { locale } = useLocale()

    let startTime = 0
    let expiredTime = 0
    let remainTime = 0
    let paused = false

    const { times, timeStart, timeStop, timeUpdate } = useCountTime(
      ({ update, stop }) => {
        remainTime = expiredTime - Date.now()

        if (remainTime > 0) {
          update(remainTime)
        } else {
          remainTime = 0
          update(remainTime)
          emit('end', {
            startTime,
            endTime: expiredTime
          })

          stop()
        }
      }
    )

    function pause() {
      if (paused) {
        return
      }

      paused = true

      emit('pause', {
        remainTime
      })

      timeStop()
    }

    function resume() {
      if (!paused) {
        return
      }

      paused = false

      emit('resume', {
        remainTime
      })

      expiredTime = remainTime + Date.now()

      timeStart()
    }

    const reset: Reset = (_timing, autoStart = true) => {
      timeStop()

      paused = !autoStart

      startTime = Date.now()
      expiredTime = _timing + startTime
      remainTime = _timing

      timeUpdate(remainTime)
      !paused && timeStart()
    }

    onMounted(() => {
      if (props.initialTiming > 0) {
        reset(props.initialTiming, !paused)
      }
    })

    expose({
      pause,
      resume,
      reset
    })

    return {
      countTime: times,
      locale,
      pause,
      resume,
      reset
    }
  }
})
</script>
