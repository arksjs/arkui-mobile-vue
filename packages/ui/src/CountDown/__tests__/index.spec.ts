import { describe, test, expect } from 'vitest'
import { shallowMount } from '@vue/test-utils'
import { CountDown } from '@/CountDown'
import { timeout } from '@arksjs/test-utils/utils'

describe('CountDown', () => {
  test.concurrent('snapshot', () => {
    const wrapper = shallowMount(CountDown, {})

    expect(wrapper.html()).toMatchSnapshot()
  })

  test.concurrent('initialTiming=50', async () => {
    const wrapper = shallowMount(CountDown, {
      props: {
        initialTiming: 50
      }
    })
    expect(wrapper.vm.countTime.time).toBe(50)

    await timeout(70)
    expect(wrapper.vm.countTime.time).toBe(0)
  })
})
