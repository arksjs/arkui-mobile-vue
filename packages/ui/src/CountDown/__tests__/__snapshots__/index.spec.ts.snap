// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CountDown > snapshot 1`] = `"<div class=\\"ta-count-down\\">0:00:00</div>"`;
